src/LsmTree.hs

```diff
@@ -13,18 +13,47 @@ module LsmTree
        , initFileSystem
        , insertTree
        , insertTable
+         -- * foldr
+       , sizeTree
+       , heightTree
        ) where
 
+import Data.Foldable (foldl')
+
 -- | Data type that stores values while program is working.
 data Tree a
-    = TreeNode (Tree a) a (Tree a)
+    = TreeNode (Tree a) !a (Tree a)
     | Leaf
     deriving stock (Eq, Show)
 
+instance Foldable Tree where
+    foldMap _ Leaf = mempty
+    foldMap f (TreeNode l x r) = foldMap f l <> f x <> foldMap f r
+        -- * inorder traverse by default
+    foldr _ ini Leaf = ini
+    foldr f ini (TreeNode l x r) = foldr f (f x (foldr f ini r)) l
+
+sizeTree :: Tree a -> Int
+sizeTree = foldl' (\len _ -> len + 1) 0
+
+foldrTree :: b -- ^ constant when leaf reached
+          -> (b -> a -> b -> b) -- ^ function when node reached
+          -> Tree a 
+          -> b
+foldrTree leaf node tree =
+    case tree of
+        Leaf -> leaf
+        (TreeNode l x r) -> node (foldT l) x (foldT r)
+  where
+    foldT = foldrTree leaf node
```

Comment:
I would recommend to add type signatures to the functions inside `where` block. This usually helps code readability.